libdrgn/program.c

```diff
@@ -788,20 +886,162 @@ struct drgn_error *drgn_program_find_prstatus_by_tid(struct drgn_program *prog,
 						     struct nstring *ret)
 {
 	struct drgn_error *err;
-	struct drgn_prstatus_map_iterator it;
-
-	assert(!(prog->flags & DRGN_PROGRAM_IS_LINUX_KERNEL));
-	err = drgn_program_cache_prstatus(prog);
+	struct drgn_thread *thread;
+	err = drgn_program_find_thread(prog, tid, &thread);
 	if (err)
 		return err;
 
-	it = drgn_prstatus_map_search(&prog->prstatus_map, &tid);
-	if (!it.entry) {
+	if (!thread) {
 		ret->str = NULL;
 		ret->len = 0;
 		return NULL;
 	}
-	*ret = it.entry->value;
+	*ret = thread->prstatus;
+	return NULL;
+}
+
+LIBDRGN_PUBLIC struct drgn_error *
+drgn_program_find_thread(struct drgn_program *prog, uint32_t tid,
+			 struct drgn_thread **ret)
+{
+	struct drgn_error *err;
+
+	if (!(prog->flags & DRGN_PROGRAM_IS_LIVE)) {
+		err = drgn_program_cache_threads(prog);
+		if (err)
+			return err;
+	}
+
+	if (prog->flags & DRGN_PROGRAM_IS_LINUX_KERNEL) {
+		*ret = malloc(sizeof(**ret));
+		if (!*ret)
+			return &drgn_enomem;
+		(*ret)->prog = prog;
+		drgn_object_init(&(*ret)->thread, prog);
+
+		struct drgn_object init_pid_ns;
+		drgn_object_init(&init_pid_ns, prog);
+		err = drgn_program_find_object(prog, "init_pid_ns", NULL,
+					       DRGN_FIND_OBJECT_VARIABLE,
+					       &init_pid_ns);
+		if (err)
+			goto kernel_out;
+		err = drgn_object_address_of(&init_pid_ns, &init_pid_ns);
+		if (err)
+			goto kernel_out;
+		err = linux_helper_find_task(&(*ret)->thread, &init_pid_ns,
+					     tid);
+		if (err)
+			goto kernel_out;
+		(*ret)->tid = tid;
+	kernel_out:
+		drgn_object_deinit(&init_pid_ns);
+		if (err) {
+			drgn_object_deinit(&(*ret)->thread);
+			free(*ret);
+		}
+		return err;
+	} else {
```

Comment:
Maybe we should return an error in the live case here indicating that it's not supported instead of silently not finding the thread.